File: dev-js-api/js/algorithms/placement/levels.js
/**
 * @fileoverview levels.js — Pure function to perform Z-axis stacking of levels and shards.
 * Ready to be ported to Rust/Python.
 */

/**
 * Pure function to perform Z-axis stacking of levels and shards.
 * Computes z_start and height for each level and updates absolute position.z of member shards.
 * Preserves local Z offset (height above level floor) for shards.
 * 
 * @param {Array<{id: number, name: string, color: string}>} levelsList - Ordered levels array
 * @param {Array} shards - Shards array
 * @param {Object} oldZStarts - Old z_starts lookup dictionary { levelId: number }
 * @returns {{levels: Array, shards: Array}} Stacked levels and shards
 */
export function layoutLevelsAndShards(levelsList, shards, oldZStarts = {}) {
  const nextLevels = JSON.parse(JSON.stringify(levelsList));
  const nextShards = JSON.parse(JSON.stringify(shards));
  
  let currentZ = 0;
  const gapBetweenLevels = 0;

  nextLevels.forEach((lvl) => {
    lvl.z_start = currentZ;

    // Find shards belonging to this level
    const lvlShards = nextShards.filter(s => s.orbit === lvl.id);

    // Auto-detect old floor if not provided in oldZStarts lookup
    let oldFloor = oldZStarts[lvl.id];
    if (oldFloor === undefined) {
      if (lvlShards.length > 0) {
        oldFloor = Math.min(...lvlShards.map(s => s.position.z));
      } else {
        oldFloor = 0;
      }
    }

    let maxLvlH = 10; // Default height if level is empty
    lvlShards.forEach(s => {
      // Calculate local Z height above level floor using the old floor position
      const localZ = Math.max(0, s.position.z - oldFloor);
      
      const shardTop = localZ + s.size.h;
      if (shardTop > maxLvlH) {
        maxLvlH = shardTop;
      }

      // Translate shard to the new absolute Z position
      s.position.z = lvl.z_start + localZ;
    });

    lvl.height = maxLvlH;
    currentZ = lvl.z_start + lvl.height + gapBetweenLevels;
  });

  return { levels: nextLevels, shards: nextShards };
}

File: dev-client-api/algorithms/placement/levels.py
#!/usr/bin/env python3
"""
levels.py — Pure Python function for stacking levels and shards along the Z-axis.
"""

import copy

def layout_levels_and_shards(levels_list, shards, old_z_starts=None):
    """
    Pure function to stack levels and shards vertically.
    Calculates z_start and height for each level, updating absolute position of shards.
    Preserves local Z height above level floor.
    """
    if old_z_starts is None:
        old_z_starts = {}

    next_levels = copy.deepcopy(levels_list)
    next_shards = copy.deepcopy(shards)

    current_z = 0

    for lvl in next_levels:
        lvl_id = lvl["id"]
        lvl["z_start"] = current_z

        # Find shards belonging to this level
        lvl_shards = [s for s in next_shards if s["orbit"] == lvl_id]

        # Auto-detect old floor if not specified in old_z_starts
        old_floor = old_z_starts.get(lvl_id)
        if old_floor is None:
            if lvl_shards:
                old_floor = min(s["position"]["z"] for s in lvl_shards)
            else:
                old_floor = 0

        max_lvl_h = 10  # Default height if level is empty
        for s in lvl_shards:
            # Calculate local Z height above level floor using the old floor position
            local_z = max(0, s["position"]["z"] - old_floor)
            
            shard_top = local_z + s["size"]["h"]
            if shard_top > max_lvl_h:
                max_lvl_h = shard_top

            # Translate shard to the new absolute Z position
            s["position"]["z"] = lvl["z_start"] + local_z

        lvl["height"] = max_lvl_h
        padding = max(0, int(lvl.get("padding", 0)))
        current_z = lvl["z_start"] + lvl["height"] + padding

    return {"levels": next_levels, "shards": next_shards}
